grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java

```diff
@@ -772,48 +772,52 @@ private void cleanUp() {
-          if (stateInfo.getState() == READY) {
-            pickList.add(entry);
-          } else if (stateInfo.getState() == TRANSIENT_FAILURE) {
-            error = stateInfo.getStatus();
-          } else {
-            hasPending = true;
-          }
-        }
-        if (pickList.isEmpty()) {
-          if (error != null && !hasPending) {
-            pickList.add(new ErrorEntry(error));
-            state = TRANSIENT_FAILURE;
-          } else {
-            pickList.add(BUFFER_ENTRY);
-            state = CONNECTING;
+    if (backendList.isEmpty()) {
+      if (balancerWorking)  {
+        pickList =
+            Collections.<RoundRobinEntry>singletonList(
+                new ErrorEntry(NO_AVAILABLE_BACKENDS_STATUS));
+        state = TRANSIENT_FAILURE;
+      } else if (usingFallbackBackends) {
+        pickList =
+            Collections.<RoundRobinEntry>singletonList(
+                new ErrorEntry(NO_FALLBACK_BACKENDS_FOUND_STATUS));
+        state = TRANSIENT_FAILURE;
+      } else {  // still waiting for balancer
+        pickList = Collections.singletonList(BUFFER_ENTRY);
+        state = CONNECTING;
+      }
+    } else {
+      switch (config.getMode()) {
+        case ROUND_ROBIN:
+          pickList = new ArrayList<>(backendList.size());
+          Status error = null;
+          boolean hasPending = false;
+          for (BackendEntry entry : backendList) {
+            Subchannel subchannel = entry.subchannel;
+            Attributes attrs = subchannel.getAttributes();
+            ConnectivityStateInfo stateInfo = attrs.get(STATE_INFO).get();
+            if (stateInfo.getState() == READY) {
+              pickList.add(entry);
+            } else if (stateInfo.getState() == TRANSIENT_FAILURE) {
+              error = stateInfo.getStatus();
+            } else {
+              hasPending = true;
+            }
           }
-        } else {
-          state = READY;
-        }
-        break;
-      case PICK_FIRST:
-        if (backendList.isEmpty()) {
-          if (lbSentEmptyBackends) {
-            pickList =
-                Collections.<RoundRobinEntry>singletonList(
-                    new ErrorEntry(NO_AVAILABLE_BACKENDS_STATUS));
-            state = TRANSIENT_FAILURE;
+          if (pickList.isEmpty()) {
+            if (hasPending) {
+              pickList.add(BUFFER_ENTRY);
+              state = CONNECTING;
+            } else {
+              pickList.add(new ErrorEntry(error));
+              state = TRANSIENT_FAILURE;
+            }
           } else {
-            pickList = Collections.singletonList(BUFFER_ENTRY);
-            // Have not received server addresses
-            state = CONNECTING;
+            state = READY;
           }
-        } else {
+          break;
+        case PICK_FIRST:
           checkState(backendList.size() == 1, "Excessive backend entries: %s", backendList);
           BackendEntry onlyEntry = backendList.get(0);
```

Comment:
Since there are local variables here, put this case in a `{}` block (normal approach is to not add any more indentation but to add the `{` on the same line as the `case PICK_FIRST:`

Reply:
ACK